pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotWindowExchangeNodeInsertRule.java

```diff
@@ -209,23 +244,37 @@ private void validateWindows(Window window) {
   }
 
   private void validateWindowAggCallsSupported(Window.Group windowGroup) {
-    for (int i = 0; i < windowGroup.aggCalls.size(); i++) {
-      Window.RexWinAggCall aggCall = windowGroup.aggCalls.get(i);
+    for (Window.RexWinAggCall aggCall : windowGroup.aggCalls) {
       SqlKind aggKind = aggCall.getKind();
       Preconditions.checkState(SUPPORTED_WINDOW_FUNCTION_KIND.contains(aggKind),
           String.format("Unsupported Window function kind: %s. Only aggregation functions are supported!", aggKind));
     }
   }
 
   private void validateWindowFrames(Window.Group windowGroup) {
-    // Has ROWS only aggregation call kind (e.g. ROW_NUMBER)?
-    boolean isRowsOnlyTypeAggregateCall = isRowsOnlyAggregationCallType(windowGroup.aggCalls);
+    // Has rank based aggregation call kind (e.g. ROW_NUMBER, RANK, DENSE_RANK)
+    boolean hasRankBasedAgg = hasRankBasedAgg(windowGroup.aggCalls);
+
+    // FOLLOWING lower bound or PRECEDING upper bound is not allowed, and won't reach here
```

Comment:
Where is this enforced?

Reply:
I tried it and iirc it is banned from parsing

Reply:
`UNBOUNDED PRECEDING` is not allowed for upper frame boundary and `UNBOUNDED FOLLOWING` is not allowed for lower frame boundary. However, `n PRECEDING` is allowed for upper frame boundary and `n FOLLOWING` is allowed for lower frame boundary (not at the same time though). This behavior is similar in both Calcite and Postgres.

Reply:
Interesting! Does it mean it can do something like `BETWEEN 5 PRECEDING AND 3 PRECEDING`?

Reply:
Yep. Postgres allows that, Calcite allows that, and with https://github.com/apache/pinot/pull/14249, Pinot will support that as well!